core/server/master/src/main/java/alluxio/master/file/UfsSyncChecker.java

```diff
@@ -119,25 +154,27 @@ public int compare(Inode a, Inode b) {
     AlluxioURI curUri = ufsUri;
     while (curUri != null) {
       if (mListedDirectories.containsKey(curUri.toString())) {
-        List<UnderFileStatus> childrenList = new LinkedList<>();
-        for (UnderFileStatus child : mListedDirectories.get(curUri.toString())) {
-          String childPath = PathUtils.concatPath(curUri, child.getName());
+        List<UfsStatus> childrenList = new LinkedList<>();
+        for (UfsStatus childStatus : mListedDirectories.get(curUri.toString())) {
+          String childPath = PathUtils.concatPath(curUri, childStatus.getName());
           String prefix = PathUtils.normalizePath(ufsUri.toString(), AlluxioURI.SEPARATOR);
           if (childPath.startsWith(prefix) && childPath.length() > prefix.length()) {
-            childrenList.add(new UnderFileStatus(childPath.substring(prefix.length()),
-                child.isDirectory()));
+            UfsStatus newStatus = childStatus.copy();
+            newStatus.setName(childPath.substring(prefix.length()));
+            childrenList.add(newStatus);
           }
         }
-        return trimIndirect(childrenList.toArray(new UnderFileStatus[childrenList.size()]));
+        return trimIndirect(childrenList.toArray(new UfsStatus[childrenList.size()]));
       }
       curUri = curUri.getParent();
     }
-    UnderFileStatus[] children =
+    UfsStatus[] children =
         ufs.listStatus(ufsUri.toString(), ListOptions.defaults().setRecursive(true));
     // Assumption: multiple mounted UFSs cannot have the same ufsUri
-    if (children != null) {
-      mListedDirectories.put(ufsUri.toString(), children);
+    if (children == null) {
+      return null;
```

Comment:
yeah, let me refactor this. `null` as a valid `listStatus` return value was not a good idea